Johnetordoff/osf.io
# [ENG-4566] Replace boto with boto3 for S3 Improvement and load S3 regions from user info

## Purpose

Upgrade to boto3 and remove old dependency of boto for general housekeeping purposes.

## Changes

Overall, the main difference between the two files lies in the underlying libraries (boto3 vs. boto) and their corresponding APIs and exception classes. Additionally, the method names and some of the parameter handling may differ between the two versions.

- Import statements:
  - Replace `from boto import exception` with `from botocore.exceptions import NoCredentialsError, ClientError`.
  - Replace `from boto.s3.connection import S3Connection` with `import boto3`.

- Connection and Client creation:
  - Replace `S3Connection` with `boto3.Session` to create a session.
  - Replace `connection = S3Connection(access_key, secret_key, calling_format=OrdinaryCallingFormat())` with `session.client('s3')` to create an S3 client.

- Bucket operations:
  - Replace `connection.get_all_buckets()` with `client.list_buckets()['Buckets']` to retrieve a list of buckets.
  - Replace `connection.create_bucket(bucket_name, location=location)` with `client.create_bucket(**create_bucket_config)` to create a bucket.
  - Replace `connection.head_bucket(bucket_name)` with `client.head_bucket(Bucket=bucket_name)` to check bucket existence.

- Exception handling:
  - Replace `exception.NoAuthHandlerFound` with `NoCredentialsError`.
  - Replace `exception.BotoServerError` with `ClientError`.
  - Update the exception handling for `ClientError` to `raise HTTPError(e.response['ResponseMetadata']['HTTPStatusCode'])`.

- Additional changes:
  - Remove the line `connection.calling_format = OrdinaryCallingFormat()`.
  - Replace `connection.get_bucket(bucket_name, validate=False).get_location()` with `client.get_bucket_location(Bucket=bucket_name)['LocationConstraint']` to retrieve bucket location.

# What going on with regions in boto?

Regions change depending on user and boto version, instead of pulling regions from a static list, lets populate them from user info via the new boto3 methods.

## What happened to the "US standard" region in boto?

In boto, the "US standard" region was represented by an empty string ('') when specifying the bucket location for the create_bucket method. This empty string indicated that the bucket would be created in the default US Standard region, which is also known as the US East (N. Virginia) region.

However, in boto3, the concept of the "US standard" region has been deprecated. The equivalent region in boto3 is us-east-1, which represents the US East (N. Virginia) region. Therefore, if you want to create a bucket in the default US Standard region using boto3, you would set the LocationConstraint parameter to us-east-1 instead of an empty string.

It's worth noting that the US East (N. Virginia) region (us-east-1) is the default region for Amazon S3 buckets. So, if you omit the LocationConstraint parameter when creating a bucket in boto3, it will be created in the US East (N. Virginia) region by default.


Please note that there may be other minor changes required depending on the specific usage and parameters.

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify new buckets have the correct regions and exsisting buckets haven't changed 
- Verify the new default region US Standard (us-east-1) is working for new and existing users. 

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Documentation



## Side Effects



## Ticket

[
](https://openscience.atlassian.net/browse/ENG-4566)
